CI/CD Pipeline for the Cyklops Events chart

alika opened the pull request:
# Context
This PR represents the work done to get chart-cyklops-events ci/cd pipeline working in GitLab CE. This work was done on a development instance of gitlab-ce (cyklopsdev cluster), but then migrated here (github.com) for CNCT engineering consistency. This migration of code is also taking place while we are pivoting from Jenkins to GitLab internally. Therefore we will need to re-verify the .gitlab-ci.yml on our official GitLab instance, when it is ready.

Fixes #5

samba commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,99 @@
+image: quay.io/samsung_cnct/helm-registry-agent:v0.7.4-helm_2.6
+
+variables:
+  KUBECONFIG: /root/.kube/config
```
Why are you storing this in the root path, and not in the working directory?
I see that you populate it later, in each job that requires it.

In particular, we need to plan for cases when CI is executing in a *different* cluster than the target Kubernetes deployment environment, and avoid cases where simultaneous jobs might operate on the same file.

samba commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,99 @@
+test:
+  stage: test
+  script:
+  - cat ${APP_NAME}/Chart.yaml
```
Why do we `cat` it here?

alika replied:
This was to simply provide better CI build logs. I used it to ensure this specific test job was testing what was built in the build steps and passed via the `artifacts` config. But it is essentially a CI debug comment and can be removed.

samba replied:
OK cool. Let's clean it out. `artifacts` is a great way to handle this, as they can also be reviewed post-hoc.

samba commented on cyklops-events/values.yaml:
```diff
@@ -0,0 +1,19 @@
+
+# Specify VictorOps service configurations
+victorops:
+  accountID: "20131114"
```
Do we need to treat this value as secret?

samba commented:
Please also have a look at [recent work on fluent-bit chart][1] too see how those are handled. If any valuable insights come from that, please feel free to integrate them here too.

[1]: https://git.cnct.io/guineveresaenger/fluent-bit/blob/gitlabci/.gitlab-ci.yml

samba commented:
@alika particularly relevant from the [other recent chart work] I referenced is the use of Git tags as version identifiers. Common Tools uses a tag-based release methodology, and we should to. 

The scripts used in that project's CI pipeline rely on Git tags to identify the current version. (There may be better ways, using [GitLab's automatic environment variables][1].)

[1]: docs.gitlab.com/ce/ci/variables/README.html#predefined-variables-environment-variables

alika commented:
Updated code base to reflect requested pipeline modifications:
- updated to git-tag based versioning
- added chart deployment test
- updated publishing jobs
- updated auto deploy task to staging

Ready for another review @samba

guineveresaenger commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,80 @@
+  script:
+  - helm lint ${APP_NAME} || exit $?
+  - helm install --name ${RELEASE} --namespace ${NAMESPACE} ./${APP_NAME}
+  - kubectl rollout status deploy/${DEPLOYMENT} -n ${NAMESPACE}
```
Instead of kubectl, can we verify successful chart deployment with a helm command?

guineveresaenger replied:
This issue might be useful? https://github.com/kubernetes/helm/issues/1520

guineveresaenger replied:
(it's a bit old unfortunately)

alika replied:
So `helm fetch --verify` looks like it retrieves and verifies the chart is built correctly. The `kubectl rollout status` command is testing if a k8s deployment has been created and is correctly running. There is a `helm status <release>` command that looks like I could parse for the same info. Using only helm commands would then drop the usage of the `kraken-tools` image all together.

samba replied:
@alika please pivot to using the `helm` command interface only.

Main considerations:

- Not all charts will produce a k8s deployment
- Reducing the dependency on kraken-tools for CI is good for performance due to its historically large volume

guineveresaenger commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,80 @@
+  - envsubst < build/Chart.yaml.in > ${APP_NAME}/Chart.yaml
+
+test:
+  image: quay.io/samsung_cnct/kraken-tools:v0.8
```
I think this should be tagged :latest. When kraken-tools breaks, all related CI runs should also fail. This is fine.

alika replied:
I think depending on `:latest` is ok when you need to see latest change to that image. In this case I just need a consistent image with `kubectl` and I want the build to operate in a immutable and repeatable manner over time. Depending on `:latest` would allow for past revisions and builds to possible behave differently over time.

But this could also be moot if we remove the need for `kubectl` and the `kraken-tools` image.

guineveresaenger replied:
I understand your point with `:latest` changing but aren't we using all these charts in combination with kraken? I may be coming at this from too much of a toolsteam perspective, but there is value to me in ensuring all charts play nicely with kraken. 
I just don't really want to go pick through all the charts and update the `kraken-tools` image to the new release. That's error-prone and hard to maintain.
It seems to me that not involving the `kraken-tools` image is a good way to avoid having to make this decision...

samba replied:
👍  for avoiding decisions ;)

Per other item below, using `helm` alone should suffice, but may still require `kubectl`. (Doesn't necessarily dictate using `kraken-tools` though.)

alika commented:
@samba was able to get a [successful build and test](https://git.cnct.io/common-tools/samsung-cnct_chart-cyklops-events/pipelines/492) of this chart on gitlab using only helm.

alika commented:
Also verified the publish and deploy jobs are [working correctly](https://git.cnct.io/common-tools/samsung-cnct_chart-cyklops-events/pipelines/496). There is an instance on the onondaga cluster.

alika commented:
After a discussion w/@samba will adapt the CD process for this chart to:

- [ ] publish to the _alpha_ channel on commits to **master**
- [ ] deploy to the **staging** environment on commits to **master** from _alpha_ channel
- [ ] publish to the _stable_ channel on well-formed version **tags**
- [ ] deploy to the **production** environment on well-formed version **tags** from _stable_ channel

alika commented:
@samba I think this is ready for review again.

samba commented on .gitlab-ci.yml:
```diff
@@ -13,14 +13,12 @@ before_script:
   stage: build
   only:
-    - master
     - branches
```
`branches` here includes `master`. Please add an `except` clause (peer with `only`) that excludes `master`, since `master` is addressed below.
